src/app/EventManagement.cpp

```diff
@@ -0,0 +1,867 @@
+
+    if (eventBuffer->IsFinalDestinationForPriority(imp))
+    {
+        // event is getting dropped.  Increase the event number and first timestamp.
+        EventNumber numEventsToDrop = 1;
+        eventBuffer->RemoveEvent(numEventsToDrop);
+        eventBuffer->mFirstEventSystemTimestamp.mValue += context.mDeltaSystemTime.mValue;
+        ChipLogProgress(EventLogging,
+                        "Dropped events from buffer with priority %d due to overflow: { event priority_level: %d, count: %d };",
+                        eventBuffer->GetPriorityLevel(), imp, numEventsToDrop);
+        ctx->mSpaceNeededForMovedEvent = 0;
+    }
+    else
+    {
+        // event is not getting dropped. Note how much space it requires, and return.
+        ctx->mSpaceNeededForMovedEvent = aReader.GetLengthRead();
+        err                            = CHIP_END_OF_TLV;
+    }
+
+exit:
+    return err;
+}
+
+CHIP_ERROR EventManagement::ScheduleFlushIfNeeded(EventOptions::Type aUrgent)
+{
+    // TODO: Implement ScheduleFlushIfNeeded
+    return CHIP_NO_ERROR;
+}
+
+void CircularEventBuffer::Init(uint8_t * apBuffer, uint32_t aBufferLength, CircularEventBuffer * apPrev,
+                               CircularEventBuffer * apNext, PriorityLevel aPriorityLevel)
+{
+    CHIPCircularTLVBuffer::Init(apBuffer, aBufferLength);
+    mpPrev            = apPrev;
+    mpNext            = apNext;
+    mPriority         = aPriorityLevel;
+    mFirstEventNumber = 1;
+    mLastEventNumber  = 0;
+    mFirstEventSystemTimestamp.Init(Timestamp::Type::kSystem, 0);
+    mLastEventSystemTimestamp.Init(Timestamp::Type::kSystem, 0);
+    mpEventNumberCounter = nullptr;
+}
+
+bool CircularEventBuffer::IsFinalDestinationForPriority(PriorityLevel aPriority) const
+{
+    return !((mpNext != nullptr) && (mpNext->mPriority <= aPriority));
+}
+
+void CircularEventBuffer::UpdateFirstLastEventTime(Timestamp aEventTimestamp)
+{
+    if (mFirstEventSystemTimestamp.mValue == 0)
+    {
+        mFirstEventSystemTimestamp = aEventTimestamp;
+        mLastEventSystemTimestamp  = aEventTimestamp;
+    }
+    mLastEventSystemTimestamp = aEventTimestamp;
+}
+
+void CircularEventBuffer::RemoveEvent(EventNumber aNumEvents)
+{
+    mFirstEventNumber = mFirstEventNumber + aNumEvents;
+}
+
+void CircularEventReader::Init(CircularEventBufferWrapper * apBufWrapper)
+{
+    TLV::TLVReader reader;
+    CircularEventBuffer * prev;
+
+    if (apBufWrapper->mpCurrent == nullptr)
+        return;
+    reader.Init(*apBufWrapper, apBufWrapper->mpCurrent->DataLength());
+    TLVReader::Init(reader);
+    mMaxLen = apBufWrapper->mpCurrent->DataLength();
+    for (prev = apBufWrapper->mpCurrent->GetPreviousCircularEventBuffer(); prev != nullptr;
+         prev = prev->GetPreviousCircularEventBuffer())
+    {
+        CircularEventBufferWrapper bufWrapper;
+        bufWrapper.mpCurrent = prev;
+        reader.Init(bufWrapper, prev->DataLength());
+        mMaxLen += reader.GetRemainingLength();
```

Comment:
How is this different from `mMaxLen += pref->DataLength()` ?

Why is the reader used in the first place?

Reply:
updated, thanks